src/runtime_src/core/edge/drm/zocl/zocl_sysfs.c

```diff
@@ -133,6 +134,60 @@ static ssize_t zocl_get_memstat(struct device *dev, char *buf, bool raw)
 	return size;
 }
 
+static ssize_t graph_status_show(struct device *dev, struct device_attribute *attr, char *buf)
+{
+	struct drm_zocl_dev *zdev;
+	size_t size;
+	struct aie_info *aie;
+	struct aie_info_cmd *acmd;
+	struct aie_info_packet *aiec_packet;
+	ssize_t nread = 0;
+	zdev = dev_get_drvdata(dev);
+	if (!zdev)
+		return 0;
+
+	aie = zdev->aie_information;
+	acmd = kmalloc(sizeof(struct aie_info_cmd), GFP_KERNEL);
+	if (!acmd) {
+		return -ENOMEM;
+	}
+
+	aiec_packet = kmalloc(sizeof(struct aie_info_packet), GFP_KERNEL);
+	if (!aiec_packet) {
+		return -ENOMEM;
+	}
+
+	aiec_packet->opcode = GRAPH_STATUS;
+	acmd->aiec_packet = aiec_packet;
+
+	/* caller release the wait aied thread and wait for result*/
+	mutex_lock(&aie->aie_lock);
+	if (waitqueue_active(&aie->aie_wait_queue)) {
+		list_add_tail(&acmd->aiec_list, &aie->aie_cmd_list);
+		mutex_unlock(&aie->aie_lock);
+		wake_up_interruptible(&aie->aie_wait_queue);
+		if (wait_event_interruptible(aie->aie_wait_result_queue,
+		    !list_empty(&aie->aie_cmd_result_list))) {
```

Comment:
Please double check to see if list_empty() is thread safe.

Reply:
@Himanshu-xilinx as we discussed off line. It is better for each sysfs node reader to wait for their own semaphore other than share one same work queue. Otherwise it will be a problem for the reader to get the result it is waiting for.

Reply:
done